Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs

```diff
@@ -261,4 +269,52 @@ private void UpdateIntercomUi(EntityUid uid, IntercomComponent component)
         var state = new IntercomBoundUIState(micEnabled, speakerEnabled, availableChannels, selectedChannel);
         _ui.TrySetUiState(uid, IntercomUiKey.Key, state);
     }
+
+    #region Handheld Radio
+
+    private void OnBeforeHandheldRadioUiOpen(Entity<RadioMicrophoneComponent> microphone, ref BeforeActivatableUIOpenEvent args)
+    {
+        UpdateHandheldRadioUi(microphone);
+    }
+
+    private void OnToggleHandheldRadioMic(Entity<RadioMicrophoneComponent> microphone, ref ToggleHandheldRadioMicMessage args)
+    {
+        if (args.Session.AttachedEntity is not { } user)
+            return;
+
+        SetMicrophoneEnabled(microphone, user, args.Enabled, true);
+        UpdateHandheldRadioUi(microphone);
+    }
+
+    private void OnToggleHandheldRadioSpeaker(Entity<RadioMicrophoneComponent> microphone, ref ToggleHandheldRadioSpeakerMessage args)
+    {
+        if (args.Session.AttachedEntity is not { } user)
+            return;
+
+        SetSpeakerEnabled(microphone, user, args.Enabled, true);
+        UpdateHandheldRadioUi(microphone);
+    }
+
+    private void OnChangeHandheldRadioFrequency(Entity<RadioMicrophoneComponent> microphone, ref SelectHandheldRadioFrequencyMessage args)
+    {
+        if (args.Session.AttachedEntity is not { })
+            return;
+
+        microphone.Comp.Frequency = args.Frequency;
+        UpdateHandheldRadioUi(microphone);
+    }
+
+    private void UpdateHandheldRadioUi(Entity<RadioMicrophoneComponent> radio)
+    {
+        var micComp = CompOrNull<RadioMicrophoneComponent>(radio);
```

Comment:
You're getting RadioMicrophone from arguments, getting micComp is redunant

Reply:
yep, you can acces to components values via `radio.Comp`